azure_data_cosmos_engine/src/hash.rs

```diff
@@ -0,0 +1,537 @@
+pub fn get_hashed_partition_key_string(
+    pk_value: &[PartitionKeyValue],
+    kind: Option<PartitionKeyKind>,
+    version: Option<u8>,
+) -> String {
+    if pk_value.is_empty() {
+        return MIN_INCLUSIVE_EFFECTIVE_PARTITION_KEY.to_string();
+    }
+    if pk_value == [PartitionKeyValue::Infinity] {
+        return MAX_EXCLUSIVE_EFFECTIVE_PARTITION_KEY.to_string();
+    }
+
+    match kind {
+        Some(PartitionKeyKind::Hash) => {
+            let version = version.unwrap_or(1);
+            match version {
+                1 => get_effective_partition_key_for_hash_partitioning_v1(pk_value),
+                2 => get_effective_partition_key_for_hash_partitioning_v2(pk_value),
+                _ => {
+                    panic!("Hash partitioning only supports version 1 or 2");
+                }
+            }
+        }
+        // hpk only supports V2
+        Some(PartitionKeyKind::MultiHash) => {
+            panic!("MultiHash currently not supported. Pending additional testing.");
+            // if version != Some(2) {
+            //     panic!("MultiHash partitioning only supports version 2");
+            // }
+            // get_effective_partition_key_for_multi_hash_partitioning_v2(pk_value)
+        }
+        _ => to_hex_encoded_binary_string(pk_value),
+    }
+}
+
+/// V2: encode components with `_write_for_hashing_v2`, hash the concatenated bytes,
+fn get_effective_partition_key_for_hash_partitioning_v2(pk_value: &[PartitionKeyValue]) -> String {
+    let mut ms: Vec<u8> = Vec::new();
+    for comp in pk_value {
+        write_for_hashing_v2(comp, &mut ms);
+    }
+    let hash_128 = murmurhash3_128(&ms, 0);
+    let mut hash_bytes = hash_128.to_le_bytes();
+    hash_bytes.reverse();
+    // Reset 2 most significant bits of first byte
+    hash_bytes[0] &= 0x3F;
+    bytes_to_hex_upper(&hash_bytes)
+}
+
+/// Multi-hash V2: compute per-component hash similarly and concatenate uppercase hex segments.
+// fn get_effective_partition_key_for_multi_hash_partitioning_v2(
+//     pk_value: &[PartitionKeyValue],
+// ) -> String {
+//     let mut pieces: Vec<String> = Vec::new();
+//     for comp in pk_value {
+//         let mut ms: Vec<u8> = Vec::new();
+//         write_for_hashing_v2(comp, &mut ms);
+//         let hash_128 = murmurhash3_128(&ms, 0);
+//         let mut hash_bytes = hash_128.to_le_bytes();
+//         hash_bytes.reverse();
+//         hash_bytes[0] &= 0x3F;
+//         pieces.push(bytes_to_hex_upper(&hash_bytes));
+//     }
+//     pieces.join("").to_uppercase()
+// }
+
+/// Truncate strings to 100 characters for V1 hashing (other types unchanged).
+fn truncate_for_v1_hashing(value: &PartitionKeyValue) -> PartitionKeyValue {
+    match value {
+        PartitionKeyValue::String(s) => {
+            if s.len() > 100 {
+                PartitionKeyValue::String(s[..100].to_string())
+            } else {
+                PartitionKeyValue::String(s.clone())
```

Comment:
We shouldn't need to clone in this case. Have `truncate_for_v1_hashing` take the `value` by value and you should be able to move the string rather than cloning it.